src/Workspaces/CoreTestUtilities/Workspaces/TestHostProject.cs

```diff
@@ -14,9 +14,125 @@
+            IEnumerable<MetadataReference> metadataReferences = null,
+            IEnumerable<AnalyzerReference> analyzerReferences = null,
+            string assemblyName = null,
+            string defaultNamespace = null)
+            : this(workspace, name, language, compilationOptions, parseOptions, [document], [], [], projectReferences, metadataReferences, analyzerReferences, assemblyName, defaultNamespace)
+        {
+        }
+
+        internal TestHostProject(
+            HostLanguageServices languageServices,
+            CompilationOptions compilationOptions,
+            ParseOptions parseOptions,
+            string assemblyName,
+            string projectName,
+            IList<MetadataReference> references,
+            IList<TestHostDocument> documents,
+            IList<TestHostDocument> additionalDocuments = null,
+            IList<TestHostDocument> analyzerConfigDocuments = null,
+            Type hostObjectType = null,
+            bool isSubmission = false,
+            string filePath = null,
+            IList<AnalyzerReference> analyzerReferences = null,
+            string defaultNamespace = null)
+            : base(languageServices,
+                   compilationOptions,
+                   parseOptions,
+                   assemblyName,
+                   projectName,
+                   references,
+                   documents,
+                   additionalDocuments,
+                   analyzerConfigDocuments,
+                   hostObjectType,
+                   isSubmission,
+                   filePath,
+                   analyzerReferences,
+                   defaultNamespace)
+        {
+        }
+
+        public TestHostProject(
+            TestWorkspace workspace,
+            string name = null,
+            string language = null,
+            CompilationOptions compilationOptions = null,
+            ParseOptions parseOptions = null,
+            IEnumerable<TestHostDocument> documents = null,
+            IEnumerable<TestHostDocument> additionalDocuments = null,
+            IEnumerable<TestHostDocument> analyzerConfigDocuments = null,
+            IEnumerable<TestHostProject> projectReferences = null,
+            IEnumerable<MetadataReference> metadataReferences = null,
+            IEnumerable<AnalyzerReference> analyzerReferences = null,
+            string assemblyName = null,
+            string defaultNamespace = null)
+            : base(workspace,
+                   name,
+                   language,
+                   compilationOptions,
+                   parseOptions,
+                   documents,
+                   additionalDocuments,
+                   analyzerConfigDocuments,
+                   projectReferences,
+                   metadataReferences,
+                   analyzerReferences,
+                   assemblyName,
+                   defaultNamespace)
+        {
+        }
+    }
+
+    public class TestHostProject<TDocument> : AbstractTestHostProject
```

Comment:
Could we rename those generic types to something like `BaseTestHostProject<T>` (and move them to a separate file) to differentiate them from the concrete workspace and editor layer types?

Reply:
Does the extra word in the name add value though? The type is already distinct by its arity. 

Then we also have `AbstractTestHostProject` non-generic base so we would have `BaseTestHostProject<>` and `AbstractTestHostProject`, which is now getting a bit confusing.

Reply:
Yea this is fine as is, I just feel it'd make the type hierarchy and their purpose clearer on a glance.

e.g. if it's up to me, I'd probabaly have

```cs
class AbstractTestHostDocument;
class TestHostDocument: AbstractTestHostDocument
class EditorTestHostDocument : AbstractTestHostDocument;

class AbstractTestHostProject;
class AbstractTestHostProject<TDocument> : AbstractTestHostProject where TDocument : AbstractTestHostDocument;
class TestHostProject : AbstractTestHostProject<TestHostDocument>;
class EditorTestHostProject : AbstractTestHostProject<EditorTestHostDocument>;

class AbstractTestHostSolution<TDocument> where TDocument : AbstractTestHostDocument;
class TestHostSolution : AbstractTestHostSolution<TestHostDocument>;
class EditorTestHostSolution : AbstractTestHostSolution<EditorTestHostDocument>;

class AbstractTestWorkspace<TDocument> : Workspace where TDocument : AbstractTestHostDocument;
class TestWorkspace : AbstractTestWorkspace<TestHostDocument>;
class EditorTestWorkspace : AbstractTestWorkspace<EditorTestHostDocument>;
```

Reply:
Tried to rename the types but IDE fails to do so :(. Will follow up.